pkg/controller/metrics/metrics_with_dynamic_labels.go

```diff
@@ -0,0 +1,120 @@
+package metrics
+
+import (
+	"sort"
+
+	"github.com/prometheus/client_golang/prometheus"
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+	"github.com/openshift/hive/apis/hive/v1/metricsconfig"
+	"github.com/openshift/hive/pkg/constants"
+)
+
+// metricsWithDynamicLabels encompasses metrics that can have optional labels.
+type metricsWithDynamicLabels interface {
+	ObserveMetricWithDynamicLabels(metav1.Object, log.FieldLogger, map[string]string, float64)
+}
+
+type DynamicLabels struct {
+	// FixedLabels just need to be a slice used for defining the metric.
+	FixedLabels []string
+	// OptionalLabels will be used both for defining the metric and by the Observe method
+	OptionalLabels map[string]string
+}
+
+type CounterVecWithDynamicLabels struct {
+	*prometheus.CounterVec
+	*DynamicLabels
+}
+
+// ObserveMetricWithDynamicLabels logs the counter metric. It simply increments the value, so last parameter isn't used.
+func (c CounterVecWithDynamicLabels) ObserveMetricWithDynamicLabels(obj metav1.Object, log log.FieldLogger,
+	fixedLabels map[string]string, _ float64) {
+	fixedLabels = verifyAllLabelsPresent(fixedLabels, c.FixedLabels)
+	curriedVec, err := c.CounterVec.CurryWith(fixedLabels)
+	// Maps cannot be sorted, however we would want to loop through them in a sorted fashion to ensure all the metric
+	// labels are added in an expected order. Get the slice of map keys first, sort them and then use it to loop over
+	// the map
+	labels := make(prometheus.Labels, len(c.OptionalLabels))
+	for _, v := range GetSortedKeys(c.OptionalLabels) {
+		labels[v] = GetLabelValue(obj, c.OptionalLabels[v])
+	}
+	curriedVec, err = curriedVec.CurryWith(labels)
+	if err != nil {
+		// The possibility of hitting an error here is low, but report it and don't observe the metric in case it happens
+		log.WithField("metric", c.CounterVec).WithError(err).Error("error observing metric")
+	} else {
+		curriedVec.WithLabelValues().Inc()
+	}
+}
+
+// verifyAllLabelsPresent ensures all fixed labels for a metric are present. If missing, it sets the corresponding label
+// to a default value
+// Note: Avoid relying on this method, as it cannot ensure the labels would get curried in the expected order
```

Comment:
I don't like adding this note - but currying accepts labels as a map which cannot be sorted in an order.  At least doing this ensures we always have all the labels (which is necessary imo - if the first ever observe is with less labels - those labels will be dropped in all the rest of the observes).
Also I ensured all of the current metric observes do not need this verification.